pkg/graveler/graveler.go

```diff
@@ -1068,39 +1068,234 @@ func (g *KVGraveler) List(ctx context.Context, repositoryID RepositoryID, ref Re
+	var preRunID string
+	var commit Commit
+	var storageNamespace StorageNamespace
+	res, err := g.db.branchLocker.MetadataUpdater(ctx, repositoryID, branchID, func() (interface{}, error) {
+		isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_COMMIT)
+		if err != nil {
+			return nil, err
+		}
+		if isProtected {
+			return nil, ErrCommitToProtectedBranch
+		}
+		repo, err := g.GetRepository(ctx, repositoryID)
+		if err != nil {
+			return "", fmt.Errorf("get repository: %w", err)
+		}
+		storageNamespace = repo.StorageNamespace
+
+		branch, err := g.db.GetBranch(ctx, repositoryID, branchID)
+		if err != nil {
+			return "", fmt.Errorf("get branch: %w", err)
+		}
+
+		// fill commit information - use for pre-commit and after adding the commit information used by commit
+		commit = NewCommit()
+
+		if params.Date != nil {
+			commit.CreationDate = time.Unix(*params.Date, 0)
+		}
+		commit.Committer = params.Committer
+		commit.Message = params.Message
+		commit.Metadata = params.Metadata
+		if branch.CommitID != "" {
+			commit.Parents = CommitParents{branch.CommitID}
+		}
+
+		if params.SourceMetaRange != nil {
+			empty, err := g.stagingEmpty(ctx, branch)
+			if err != nil {
+				return nil, fmt.Errorf("checking empty branch: %w", err)
+			}
+			if !empty {
+				return nil, ErrCommitMetaRangeDirtyBranch
+			}
+		}
+
+		preRunID = g.hooks.NewRunID()
+		err = g.hooks.PreCommitHook(ctx, HookRecord{
+			RunID:            preRunID,
+			EventType:        EventTypePreCommit,
+			SourceRef:        branchID.Ref(),
+			RepositoryID:     repositoryID,
+			StorageNamespace: storageNamespace,
+			BranchID:         branchID,
+			Commit:           commit,
+		})
+		if err != nil {
+			return "", &HookAbortError{
+				EventType: EventTypePreCommit,
+				RunID:     preRunID,
+				Err:       err,
+			}
+		}
+
+		var branchMetaRangeID MetaRangeID
+		var parentGeneration int
+		if branch.CommitID != "" {
+			branchCommit, err := g.RefManager.GetCommit(ctx, repositoryID, branch.CommitID)
+			if err != nil {
+				return "", fmt.Errorf("get commit: %w", err)
+			}
+			branchMetaRangeID = branchCommit.MetaRangeID
+			parentGeneration = branchCommit.Generation
+		}
+		commit.Generation = parentGeneration + 1
+		if params.SourceMetaRange != nil {
+			commit.MetaRangeID = *params.SourceMetaRange
+		} else {
+			changes, err := g.StagingManager.List(ctx, branch.StagingToken, ListingMaxBatchSize)
+			if err != nil {
+				return "", fmt.Errorf("staging list: %w", err)
+			}
+			defer changes.Close()
+
+			commit.MetaRangeID, _, err = g.CommittedManager.Commit(ctx, storageNamespace, branchMetaRangeID, changes)
+			if err != nil {
+				return "", fmt.Errorf("commit: %w", err)
+			}
+		}
+
+		// add commit
+		newCommit, err := g.RefManager.AddCommit(ctx, repositoryID, commit)
+		if err != nil {
+			return "", fmt.Errorf("add commit: %w", err)
+		}
+		err = g.RefManager.SetBranch(ctx, repositoryID, branchID, Branch{
```

Comment:
In the new flow, one of the first things we should do before creating the new commit, is moving the staging token to the sealed tokens and setting a new staging token